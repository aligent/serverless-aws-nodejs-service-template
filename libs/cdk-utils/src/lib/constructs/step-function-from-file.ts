import type { Function } from 'aws-cdk-lib/aws-lambda';
import {
    CfnStateMachineAlias,
    CfnStateMachineVersion,
    DefinitionBody,
    StateMachine,
    type StateMachineProps,
} from 'aws-cdk-lib/aws-stepfunctions';
import { propertyInjectable } from 'aws-cdk-lib/core/lib/prop-injectable';
import { Construct } from 'constructs';

export interface StepFunctionFromFileProps extends StateMachineProps {
    readonly filepath: string;
    readonly alias?: string;
    readonly lambdaFunctions?: Function[];
}

/**
 * Narrowed type helper for setting global defaults
 *
 * Exclusions:
 * filepath, definitionSubstitutions, lambdaFunctions - doesn't make sense to set on a global level
 */
type StepFunctionFromFileContext = Omit<
    StepFunctionFromFileProps,
    'filepath' | 'definitionSubstitutions' | 'lambdaFunctions'
>;

/**
 * Merge in definition substitutions supporting the format `Resource: ${lambdaId}`
 * in YAML syntax
 *
 * @param definitionSubstitutions - The definition substitutions to merge in.
 * @param lambdaFunctions - The lambda functions to merge in.
 * @returns The merged definition substitutions in an object that can be spread in to StepFunctionProps
 */
function prepareDefinitionSubstitutionsObject(
    definitionSubstitutions: Record<string, string> | undefined,
    lambdaFunctions: Function[] | undefined
) {
    if (!lambdaFunctions?.length) {
        return definitionSubstitutions ? { definitionSubstitutions } : {};
    }

    const lambdaDefinitions = Object.fromEntries(
        lambdaFunctions.map(fn => [fn.node.id, fn.functionArn])
    );

    return { definitionSubstitutions: { ...definitionSubstitutions, ...lambdaDefinitions } };
}

@propertyInjectable
export class StepFunctionFromFile extends StateMachine {
    readonly PROPERTY_INJECTION_ID = '@aligent.cdk-utils.StepFunctionFromFile';
    static readonly CONTEXT_KEY = '@aligent.cdk-utils.StepFunctionFromFile';

    static defineContext(props: StepFunctionFromFileContext) {
        return {
            [this.CONTEXT_KEY]: props,
        };
    }

    /**
     * Get the context for the StepFunctionFromFile construct.
     *
     * @param scope - The scope to get the context from.
     * @returns The context for the StepFunctionFromFile construct.
     */
    static getContext(scope: Construct): StepFunctionFromFileContext {
        return scope.node.tryGetContext(StepFunctionFromFile.CONTEXT_KEY) || {};
    }

    constructor(scope: Construct, id: string, props: StepFunctionFromFileProps) {
        const defaults = StepFunctionFromFile.getContext(scope);

        const definitionSubstitutionsObject = prepareDefinitionSubstitutionsObject(
            props.definitionSubstitutions,
            props.lambdaFunctions
        );

        const { filepath, alias, ...newProps } = {
            ...defaults,
            ...props,
            ...definitionSubstitutionsObject,
        };

        super(scope, id, {
            definitionBody: DefinitionBody.fromFile(props.filepath),
            ...newProps,
        });

        // If lambda functions are provided, give the state machine
        // permission to invoke them
        // TODO Is there a more efficient way to do this?
        if (props.lambdaFunctions) {
            props.lambdaFunctions.forEach(fn => {
                fn.grantInvoke(this);
            });
        }

        if (alias) {
            const version = new CfnStateMachineVersion(this, `Version`, {
                stateMachineArn: this.stateMachineArn,
            });

            new CfnStateMachineAlias(this, `Alias`, {
                name: alias,
                routingConfiguration: [
                    {
                        stateMachineVersionArn: version.attrArn,
                        weight: 100,
                    },
                ],
            });
        }
    }
}
