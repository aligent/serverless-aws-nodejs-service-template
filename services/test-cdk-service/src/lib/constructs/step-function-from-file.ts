import type { Stack } from 'aws-cdk-lib';
import {
    CfnStateMachineAlias,
    CfnStateMachineVersion,
    DefinitionBody,
    StateMachine,
    type StateMachineProps,
} from 'aws-cdk-lib/aws-stepfunctions';

export interface StepFunctionFromFileProps extends StateMachineProps {
    readonly filepath: string;
}

export class StepFunctionFromFile extends StateMachine {
    constructor(scope: Stack, id: string, props: StepFunctionFromFileProps) {
        super(scope, id, {
            definitionBody: DefinitionBody.fromFile(props.filepath),
            // TODO: This is a hack to get the stack name into the state machine
            definitionSubstitutions: {
                stackName: scope.stackName,
            },
            tracingEnabled: true,
            stateMachineName: `${scope.stackName}-${id}`,
            role: props.role,
            ...props,
        });

        // const sentenceCaseId = `${id.charAt(0).toUpperCase()}${id.slice(1)}`;

        const version = new CfnStateMachineVersion(this, `Version`, {
            stateMachineArn: this.stateMachineArn,
        });

        new CfnStateMachineAlias(this, `Alias`, {
            name: 'LATEST',
            routingConfiguration: [
                {
                    stateMachineVersionArn: version.attrArn,
                    weight: 100,
                },
            ],
        });
    }
}
