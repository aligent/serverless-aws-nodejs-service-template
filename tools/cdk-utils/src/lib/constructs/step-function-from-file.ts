import {
    CfnStateMachineAlias,
    CfnStateMachineVersion,
    DefinitionBody,
    StateMachine,
    type StateMachineProps,
} from 'aws-cdk-lib/aws-stepfunctions';
import { propertyInjectable } from 'aws-cdk-lib/core/lib/prop-injectable';
import { Construct } from 'constructs';

export interface StepFunctionFromFileProps extends StateMachineProps {
    readonly filepath: string;
    readonly alias?: string;
}

@propertyInjectable
export class StepFunctionFromFile extends StateMachine {
    readonly PROPERTY_INJECTION_ID = '@aligent.cdk-utils.StepFunctionFromFile';
    static readonly CONTEXT_KEY = '@aligent.cdk-utils.StepFunctionFromFile';

    static defineContext(props: Omit<StepFunctionFromFileProps, 'filepath'>) {
        return {
            [this.CONTEXT_KEY]: props,
        };
    }

    constructor(scope: Construct, id: string, props: StepFunctionFromFileProps) {
        const defaults = scope.node.tryGetContext(StepFunctionFromFile.CONTEXT_KEY) || {};

        super(scope, id, {
            definitionBody: DefinitionBody.fromFile(props.filepath),
            ...defaults,
            ...props,
        });

        if (props.alias) {
            const version = new CfnStateMachineVersion(this, `Version`, {
                stateMachineArn: this.stateMachineArn,
            });

            new CfnStateMachineAlias(this, `Alias`, {
                name: props.alias,
                routingConfiguration: [
                    {
                        stateMachineVersionArn: version.attrArn,
                        weight: 100,
                    },
                ],
            });
        }
    }
}
