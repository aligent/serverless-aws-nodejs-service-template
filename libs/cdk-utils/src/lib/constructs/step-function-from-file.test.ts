import { Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { join as pathJoin } from 'path';
import { beforeEach, describe, expect, test } from 'vitest';
import { StepFunctionFromFile } from './step-function-from-file';

const snapshotMessage = 'Rerun tests with the -u flag to update snapshots if changes are expected';

describe('StepFunctionFromFile', () => {
    let stack: Stack;

    beforeEach(() => {
        stack = new Stack();
        new StepFunctionFromFile(stack, 'MyStateMachine', {
            filepath: pathJoin(__dirname, '__data__', 'test-machine.asl.yaml'),
        });
    });

    test('creates a state machine from a file', () => {
        const template = Template.fromStack(stack);

        const stateMachines = template.findResources('AWS::StepFunctions::StateMachine');
        expect(stateMachines).toMatchSnapshot(snapshotMessage);
    });

    test('creates a state machine with lambda functions', () => {
        // Create test lambda functions
        const lambda = new NodejsFunction(stack, 'TestFunction', {
            runtime: Runtime.NODEJS_22_X,
            handler: 'index.handler',
            code: Code.fromInline('exports.handler = async () => ({ statusCode: 200 });'),
        });

        const otherLambda = new NodejsFunction(stack, 'OtherTestFunction', {
            runtime: Runtime.NODEJS_22_X,
            handler: 'index.handler',
            code: Code.fromInline('exports.handler = async () => ({ statusCode: 200 });'),
        });

        new StepFunctionFromFile(stack, 'LambdaStateMachine', {
            filepath: pathJoin(__dirname, '__data__', 'test-machine.asl.yaml'),
            lambdaFunctions: [lambda, otherLambda],
            definitionSubstitutions: {
                ExtraParam: 'ExtraValue',
            },
        });

        const template = Template.fromStack(stack);

        template.hasResourceProperties('AWS::StepFunctions::StateMachine', {
            DefinitionSubstitutions: {
                ExtraParam: 'ExtraValue',
                TestFunction: { 'Fn::GetAtt': [Match.stringLikeRegexp('TestFunction.*'), 'Arn'] },
                OtherTestFunction: {
                    'Fn::GetAtt': [Match.stringLikeRegexp('OtherTestFunction.*'), 'Arn'],
                },
            },
        });

        // Check that IAM policies grant invoke permissions (there should be 2 policies, one for each lambda)
        template.hasResourceProperties('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: Match.arrayWith([
                    Match.objectLike({
                        Action: 'lambda:InvokeFunction',
                        Effect: 'Allow',
                    }),
                ]),
            },
        });
    });
});
