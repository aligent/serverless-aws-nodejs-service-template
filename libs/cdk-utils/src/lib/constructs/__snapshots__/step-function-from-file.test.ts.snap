// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StepFunctionFromFile > integration tests > creates a state machine from a file 1`] = `
{
  "MyStateMachine6C968CA5": {
    "DeletionPolicy": "Delete",
    "DependsOn": [
      "MyStateMachineRoleD59FFEBC",
    ],
    "Properties": {
      "DefinitionS3Location": {
        "Bucket": "[SNAPSHOT_PLACEHOLDER]",
        "Key": "[SNAPSHOT_PLACEHOLDER]",
      },
      "RoleArn": {
        "Fn::GetAtt": [
          "MyStateMachineRoleD59FFEBC",
          "Arn",
        ],
      },
    },
    "Type": "AWS::StepFunctions::StateMachine",
    "UpdateReplacePolicy": "Delete",
  },
}
`;
